posts/2022-09-23_flight/index.qmd

```diff
@@ -0,0 +1,821 @@
+
+```{python, attr.source='.numberLines startFrom="7"'}
+#| filename: "tiny_flight.py [server]"
+#| code-line-numbers: true
+#| eval: false
+class TinyServer(flight.FlightServerBase):
+  
+    def __init__(self, 
+                 host = 'localhost', 
+                 port = 5678):
+        self.tables = {}
+        self.location = flight                  \
+                        .Location               \
+                        .for_grpc_tcp(host, port)
+        super().__init__(self.location)
+```
+
+```{python, attr.source='.numberLines startFrom="96"'}
+#| filename: "tiny_flight.py [client]"
+#| code-line-numbers: true
+#| eval: false
+class TinyClient:
+
+    def __init__(self, host = 'localhost', port = 5678):
+        self.location = flight                      \
+                        .Location                   \
+                        .for_grpc_tcp(host, port)
+        self.connection = flight.connect(self.location)
+        self.connection.wait_for_available()
+```
+
+Some things to notice here. At start up, the server and client both call the `flight.Location.for_grpc_tcp()` function to generate a  [Location](https://arrow.apache.org/docs/python/generated/pyarrow.flight.Location.html) object used to specify the address of the server:
+
+```{python}
+loc = flight.Location.for_grpc_tcp('localhost', 5678)
+print(loc)
+```
+
+The important thing in this output the server address. The `localhost:5678` part indicates that the server is running locally on port 5678, and the `grpc+tcp://` part tells us what communication protocols are being used. For this server, those protocols are [gRPC](https://grpc.io/) and [TCP](https://en.wikipedia.org/wiki/Transmission_Control_Protocol). TCP is probably familiar to most data scientists since it's one of the core protocols of the internet, but gRPC (wonderful as it is) is a little more specialised. I'm not going to talk about how gRPC works in this post, but there are some references at the end. For now, it's sufficient to recognise that this location object does store the server address. If I'd really wanted to, I could have written code that constructs this string manually^[People with more experience in Python (well, more than me) would notice the importance of the `b''` notation. One thing I didn't know about Python until recently is that it is quite explicit in specifying how strings are encoded. The `b''` notation is used to indicate that this is a "byte literal" string. To convert it to utf-8 text, it needs to be explicitly decoded. I mention this here because later on in the post I'm going to call `.encode()` and `.decode()` string methods to switch back and forth between byte literals and utf-8 strings. I'm assuming this is common knowledge among Python users, but coming from R this was a little surprising!] but there's no need to do that when the pyarrow flight module supplies built-in location classes to do this for us!
```

Comment:
```suggestion
The important thing in this output is the server address. The `localhost:5678` part indicates that the server is running locally on port 5678, and the `grpc+tcp://` part tells us what communication protocols are being used. For this server, those protocols are [gRPC](https://grpc.io/) and [TCP](https://en.wikipedia.org/wiki/Transmission_Control_Protocol). TCP is probably familiar to most data scientists since it's one of the core protocols of the internet, but gRPC (wonderful as it is) is a little more specialised. I'm not going to talk about how gRPC works in this post, but there are some references at the end. For now, it's sufficient to recognise that this location object does store the server address. If I'd really wanted to, I could have written code that constructs this string manually^[People with more experience in Python (well, more than me) would notice the importance of the `b''` notation. One thing I didn't know about Python until recently is that it is quite explicit in specifying how strings are encoded. The `b''` notation is used to indicate that this is a "byte literal" string. To convert it to utf-8 text, it needs to be explicitly decoded. I mention this here because later on in the post I'm going to call `.encode()` and `.decode()` string methods to switch back and forth between byte literals and utf-8 strings. I'm assuming this is common knowledge among Python users, but coming from R this was a little surprising!] but there's no need to do that when the pyarrow flight module supplies built-in location classes to do this for us!
```